### DISCUSSION TITLE
Decorations

### DISCUSSION THREAD

**Discussion by stephengoldbaum** [finos/morphir#136] [Ideas]
# Decorations Intro
The Morphir project has had many requires for the ability to annotate specific items in a Morphir model with information beyond that provided by the IR.  Many languages feature such a capability through mechanisms like annotations or decorations.  This discussion seeks to collect the set of feature requirements for adding decorations to Morphir.  To start we, have collected the following features from the various requests:

# User Stories
1. As a Morphir backend developer, I would like to **annotate certain constructs with extra structured information** to provide my backend processor with necessary information that isn't available in the IR.
2. As a Morphir backend developer, I would like a standard way to retrieve relevant decoration values so that I am insulated from the underlying decoration implementation.  Furthermore, this should include the ability to retrieve all decoration values for a given item and the ability to find all items associated with a particular decoration type.
3. As a Morphir backend developer, I would like a process to add my decoration requirements to the Morphir tooling.
4. As a Morphir author, I would like a way to choose what backends/decorations are used in my workspace.
5.  As a Morphir author, I would like to know the structure of the decorations required by the backends that I'm using.
6. As a Morphir author, I would like a way to edit the decoration values for my Morphir app in order to use my desired backend tools.
7. As a Morphir author, I would like a way to edit the decoration values for items in imported Morphir libraries in order to comply with backend tools requirements.
8. As a Morphir author, I would like my decoration values to be packaged and distributed along with my Morphir IR.
9. As a Morphir author, I would like to know when the decorations I use are conceptually in conflict so that I know how to handle that situation in my backend. 

# Resulting Questions
The list of users leads to the following questions that must be answered by any proposal:
1. **What gets decorated?**  Just types, modules, and functions or anything?
2. **Where and how do we edit decoration values?**
3. **How do we access decoration values from our backend tools?**
4. **How do we define new decorations and integrate them into the Morphir tooling?**
5. **How do we include decorations into our workspace?**
6. **How are decorations stored and distributed with the IR?**
7. **How are items being decorated uniquely identified across languages?**

Once this set of requirements is signed off, we'll submit a proposed specification and solution.

**Comment by AttilaMihaly** [comment]
As we were discussing potential solutions we realized that it would worth further clarifying what "a given node" means in story 2. Based on the description it might sound like you can take a type or a value, pass it to some tool or function and get the decorations back. But since types and values in the Morphir IR are referentially transparent, that would mean that you cannot assign different decorations to the same type even if it's used in different parts of the model. 

So when decorating types and values we need to use an identifier that refers to the location of the type or value within the IR, such as a fully-qualified name. Since this was a bit abstract let me represent with an example:

```elm
type Tide = High | Low

type Volume = High | Low
```

The type here is `High | Low`. `Tide` and `Volume` are the names of those types. If we assigned decorations to types we could only assign the same decorations to `Tide` and `Volume`. To achieve what we want we need to assign decorations to their names. 

Not sure if we want to change the story itself to reflect this or add it as a note, but I think it's important to clarify to avoid any confusion.

**Reply by stephengoldbaum** [reply]
Changed node to item to be less technical

**Comment by DamianReeves** [comment]
Could we add:

- As a Morphir backend developer, I would like a way to retrieve a single file/document where all decorations are applied to the IR/model

**Reply by stephengoldbaum** [reply]
I added question #7 to take this into account. Basically, how do we uniquely identify an item?

**Reply by AttilaMihaly** [reply]
I could be wrong but I think @DamianReeves meant something more specific: a way to attach the decorations that are stored separately to the IR directly as an attribute so that it is easy to process in a backend.

**Comment by stephengoldbaum** [comment]
[New Decorations Proposal](https://github.com/finos/morphir/discussions/141) created.

**Comment by DamianReeves** [comment]
Requesting:
1. As a Morphir tooling developer we would like Morphir to provide a specification on how merging of decorators with the IR should occur, explicitly with an eye towards collisions in node values. 
2. As a Morphir tooling develop I would like Morphir to provide tooling that merges 1 or more decorator files with a Morphir IR file, so that other tools can work with this merged file.

**Reply by stephengoldbaum** [reply]
Is this a accurate summation?

- As a Morphir tooling developer, I would like the ability to predictably merge all individual decoration units into a single combined set so that tools can work with multiple decoration types.

**Comment by DamianReeves** [comment]
I think this example of the above from smithy might help: https://smithy.io/2.0/spec/model.html#merging-models

**Reply by AttilaMihaly** [reply]
Most of that section doesn't apply to Morphir since it talks about merging models which is not supported in Morphir. It does mention merging metadata which could be applied to decorations and is detailed in this section: https://smithy.io/2.0/spec/model.html#merging-metadata.

The rules they laid out are simple enough to consider as a possible approach for decorations but as @stephengoldbaum pointed out we are getting into implementation details. In terms of requirements, I think his user story above captures the fundamental requirement, right?
